x/wasm/types/authz.go

```diff
@@ -16,12 +18,92 @@ import (
 const gasDeserializationCostPerByte = uint64(1)
 
 var (
+	_ authztypes.Authorization         = &StoreCodeAuthorization{}
 	_ authztypes.Authorization         = &ContractExecutionAuthorization{}
 	_ authztypes.Authorization         = &ContractMigrationAuthorization{}
 	_ cdctypes.UnpackInterfacesMessage = &ContractExecutionAuthorization{}
 	_ cdctypes.UnpackInterfacesMessage = &ContractMigrationAuthorization{}
 )
 
+// NewStoreCodeAuthorization constructor
+func NewStoreCodeAuthorization(grants ...CodeGrant) *StoreCodeAuthorization {
+	return &StoreCodeAuthorization{
+		Grants: grants,
+	}
+}
+
+// MsgTypeURL implements Authorization.MsgTypeURL.
+func (a StoreCodeAuthorization) MsgTypeURL() string {
+	return sdk.MsgTypeURL(&MsgStoreCode{})
+}
+
+// Accept implements Authorization.Accept.
+func (a *StoreCodeAuthorization) Accept(ctx sdk.Context, msg sdk.Msg) (authztypes.AcceptResponse, error) {
+	var code []byte
+	var permission AccessConfig
+	switch msg := msg.(type) {
+	case *MsgStoreCode:
+		code = msg.WASMByteCode
+		permission = *msg.InstantiatePermission
+	default:
+		return authztypes.AcceptResponse{}, sdkerrors.ErrInvalidRequest.Wrap("unknown msg type")
+	}
+
+	checksum, err := wasmvm.CreateChecksum(code)
+	if err != nil {
+		return authztypes.AcceptResponse{}, sdkerrors.ErrInvalidRequest.Wrap("checksum")
+	}
+
+	for _, grant := range a.Grants {
+		if grant.Accept(checksum, permission) {
+			return authztypes.AcceptResponse{Accept: true}, nil
+		}
+	}
+	return authztypes.AcceptResponse{Accept: false}, nil
+}
+
+// ValidateBasic implements Authorization.ValidateBasic.
+func (a StoreCodeAuthorization) ValidateBasic() error {
+	if len(a.Grants) == 0 {
+		return ErrEmpty.Wrap("grants")
+	}
+	for i, v := range a.Grants {
+		if err := v.ValidateBasic(); err != nil {
+			return errorsmod.Wrapf(err, "position %d", i)
+		}
+	}
+	return nil
+}
+
+// NewCodeGrant constructor
+func NewCodeGrant(codeHash []byte, instantiatePermission AccessConfig) (*CodeGrant, error) {
+	return &CodeGrant{
+		CodeHash:              codeHash,
+		InstantiatePermission: &instantiatePermission,
+	}, nil
+}
+
+// ValidateBasic validates the grant
+func (g CodeGrant) ValidateBasic() error {
+	if len(g.CodeHash) == 0 {
+		return ErrEmpty.Wrap("code hash")
+	}
+	if g.InstantiatePermission != nil {
+		if err := g.InstantiatePermission.ValidateBasic(); err != nil {
+			return errorsmod.Wrap(err, "instantiate permission")
+		}
+	}
+	return nil
+}
+
+// Accept checks if checksum and permission match the grant
+func (g CodeGrant) Accept(checksum []byte, permission AccessConfig) bool {
+	if !bytes.Equal(g.CodeHash, []byte("*")) && !bytes.Equal(g.CodeHash, checksum) {
```

Comment:
pp: If you would extract `*` to a constant with verbose naming like: `codehashWildcard`, then it would be more readable.  Just place it before the `if` if you don't to make it global